### ISSUE TITLE
[Docs] Anchor parameter is stripped out of redirected links

### ISSUE THREAD

**Issue by aswanson-nr** [newrelic/docs-website#2139]
### Related Gatsby Issue

https://github.com/gatsbyjs/gatsby/discussions/27582

### Release Date

> We merged gatsbyjs/gatsby#32334 and released it in `gatsby@next` so you could try it out there. It'll then be released on July 20th in 3.10

### Description

The redirects generated by the `gatsby-plugin-meta-redirect` plugin do not remove the trailing slash. This causes any `#anchor` parameters to be stripped out of the link and most likely applies to query parameters as well.

### Steps to reproduce
1. Go to https://docs.newrelic.com/docs/insights/new-relic-insights/using-new-relic-query-language/nrql-reference#functions)[https://docs.newrelic.com/docs/insights/new-relic-insights/using-new-relic-query-language/nrql-reference#functions
2. Notice that the page redirects
3. Notice that the `#functions` was removed from the URL and the page did not scroll down to that section

### Expected 

When a user clicks on the link in the example above, they should land on the Functions section of the page.

<img width="956" alt="Screen Shot 2021-05-06 at 9 25 51 AM" src="https://user-images.githubusercontent.com/1395158/117324562-26b45780-ae4d-11eb-947c-1e8a8ee7364f.png">

### Other information
* @jerelmiller submitted an issue to the plugin we use to generate these redirects to add an option to remove trailing slashes https://github.com/nsresulta/gatsby-plugin-meta-redirect/issues/11

### Acceptance Criteria

- [x] Test using Gastby next release 
- [x] If tests pass, upgrade to `Gatsby 3.10` when available.
- [x] Upgrade Developer SIte
- [ ] Merge Developer `develop` branch in to `feature/o11y-packs`
- [ ] Upgrade OSS site
- [ ] Upgrade private docs
- [ ] Upgrade Theme

**Comment by austin-schaefer** [comment]
Thanks @aswanson-nr . There are a lot of redirects on the site and a lot of anchor links, so a fix here will be very helpful!

**Comment by jpvajda** [comment]
from @zstix  I believe those redirects come from https://www.gatsbyjs.com/docs/reference/config-files/actions/#createRedirect in Gatsby. Not sure if we have the option to pass along the extra URL params

**Comment by rudouglas** [comment]
Notes:
Not sure if this is an issue anymore, i've having trouble reproducing it. 

https://docs.newrelic.com/docs/insights/new-relic-insights/using-new-relic-query-language/nrql-reference#functions
Redirects to 
https://docs.newrelic.com/docs/query-your-data/nrql-new-relic-query-language/get-started/nrql-syntax-clauses-functions/

Example Anchors that redirect correctly:

- https://docs.newrelic.com/docs/query-your-data/nrql-new-relic-query-language/get-started/nrql-syntax-clauses-functions/#state-select
- https://docs.newrelic.com/docs/query-your-data/nrql-new-relic-query-language/get-started/introduction-nrql-new-relics-query-language/#start-learning
- https://docs.newrelic.com/docs/telemetry-data-platform/ingest-manage-data/get-started/get-know-telemetry-data-platform/#capabilities
- https://docs.newrelic.com/docs/full-stack-observability/observe-everything/get-started/get-started-full-stack-observability/#start-anywhere
- https://docs.newrelic.com/docs/alerts-applied-intelligence/new-relic-alerts/get-started/introduction-alerts/#security

Tested on Chrome/Firefox/Safari and all seem to be anchoring correctly

**Comment by nr-kkenney** [comment]
@jpvajda it looks like we can't implement this easily since gatsby doesn't have support for this currently (see [here](https://github.com/gatsbyjs/gatsby/discussions/27582)).

is it cool if close this ticket out?

**Comment by jpvajda** [comment]
I'm going to bump Gatsby about it, and try to get an ETA of when this may be available in Gatsby.

**Comment by pieh** [comment]
Hey @jpvajda! We published today new release (`gatsby@3.5.1`), I'm currently veryfying it and will open PR with bump if all things look good.

**Comment by pieh** [comment]
I want to correct myself - I meant to write this to different (but still related to redirects) issue ( https://github.com/newrelic/docs-website/issues/1874 ). For forwarding query params and hash, we didn't implement that in Gatsby yet

**Comment by rudouglas** [comment]
no worries at all @pieh 😸

**Comment by nr-kkenney** [comment]
@rudouglas just to give you an update. had a quick chat with @roadlittledawn near the end of the day about this ticket just to get his thoughts, and he was thinking this might be difficult. 

in particular, his question was how do we differentiate a hash in a url we know is a redirect versus a hash in a url from exploring a page, since not every hash needs to be carried forward to subsequent pages viewed.

so before you go too far, might double check that case.

**Comment by rudouglas** [comment]
@nr-kkenney ye I had considered that, I was storing the anchor, and then only applying it to the path if there wasn't already an anchor after the route change. So it was only catching instances where the anchor was being stripped. After more testing it doesn't look like this is going to be possible without an update on the Gatsby side to the `createRedirect` function.

I tested every Browser API available in Gatsby and none of them catch the anchor from the path before it redirects 😞  https://gatsbyjs-org-prefetched.netlify.app/docs/browser-apis

I think we should move this to blocked but lets chat and decide today

**Comment by jpvajda** [comment]
Our decision is to hold off on trying to solve this and wait for Gatsby to fix in place for it, I'll follow up with them on an ETA.

**Comment by jpvajda** [comment]
Adding to sprint 7 so I can follow up on this issue

**Comment by jpvajda** [comment]
waiting on confirmation from gatsby this fix has been released gatsbyjs/gatsby#27582

**Comment by jpvajda** [comment]
Gatsby said this should be released in July, so once it's available it may just require us to  bump the gatsby version when it's released

**Comment by LekoArts** [comment]
We merged https://github.com/gatsbyjs/gatsby/pull/32334 and released it in `gatsby@next` so you could try it out there. It'll then be released on July 20th in 3.10

**Comment by rudouglas** [comment]
@LekoArts that's awesome thanks, I tested this locally and seems to work great, currently building a preview on this PR to test it builds ok:

https://github.com/newrelic/docs-website/pull/3160

**Comment by jpvajda** [comment]
@rudouglas Thanks for testing this out!  My thought is we wait until this feature is in a production version of Gatsby and is officially released before upgrading. it looks like it's targeted for `Gatsby 3.11`

**Comment by LekoArts** [comment]
It‘ll be in 3.10 tomorrow :)

**Comment by jpvajda** [comment]
@rudouglas to mark this done, can we just bump all sites to  Gatsby v 3.10? We don't do redirects on Dev or OSS, but it's nice to keep Gatsby version in sync IMHO.

**Comment by rudouglas** [comment]
Developer: https://github.com/newrelic/developer-website/pull/1495
OSS: https://github.com/newrelic/opensource-website/pull/818
Beta: https://github.com/newrelic/beta-docs-site/pull/93
Theme:

**Comment by jpvajda** [comment]
In our standup we discussed this is not working as expected on the Docs site after the theme bump, so @rudouglas what next steps are you taking for this issue?

**Comment by rudouglas** [comment]
@LekoArts we have been doing some testing on this and the issue does not actually seem to be resolved. You can see the behaviour via this link. We upgraded to `Gatsby v3.10.1` but the same behaviour persists both locally and in Gatsby Cloud:

https://docswebsitedevelop-rubumpgatsby.gtsb.io/docs/insights/new-relic-insights/using-new-relic-query-language/nrql-reference#functions

You will see the hash parameter does not persist when the page redirects. Would you mind taking a look and see if there could be something else causing this issue on your end? Cheers :D

**Comment by LekoArts** [comment]
Aye, thanks for the note. As a start I've added the e2e tests that were missing on my previous PR and they also show this. Will work on a fix for this :)

**Comment by jpvajda** [comment]
@LekoArts thanks for pointing that out! We'll look into the recommendations made by @KyleAMathews .

**Comment by LekoArts** [comment]
I've published https://github.com/gatsbyjs/gatsby/pull/32850 under `gatsby@next` so that you can try it out in a PR build to see if it's fixed.

I don't think that you'll need to bump the dependency in your theme as it's only a `devDependency`. Depending on how you actually use your theme you can drop the `devDependencies` all together. You'll only need this if you want to run your theme standalone without a demo/example project that uses the theme and has all the dependencies installed that the theme needs.

So for example I couldn't run https://github.com/LekoArts/gatsby-themes/tree/master/themes/gatsby-theme-cara standalone with `gatsby develop`, it wouldn't work. I need to run `gatsby develop` in https://github.com/LekoArts/gatsby-themes/tree/master/examples/cara to have a preview of the theme.

**Comment by rudouglas** [comment]
Excellent thanks @LekoArts, i'll do some testing on this and keep you udpated 🥳 

Currently building with some test changes on `gatsby@next` : #3555

**Comment by rudouglas** [comment]
This works nicely in `develop` locally...

**Comment by LekoArts** [comment]
> Excellent thanks @LekoArts, i'll do some testing on this and keep you udpated 🥳
> 
> Currently building with some test changes on `gatsby@next` : #3555

I think for I'd propose to first just upgrade `gatsby` to `next` (which I don't see in the PR you linked) and then when it works trying out turning off the `redirectInBrowser` :)

**Comment by rudouglas** [comment]
Ah yes i messed up the original PR, but this still doesn't seem to be working after upgrading to next, am I missing something @LekoArts 

#3574 
https://docswebsitedevelop-ruparameterfix.gtsb.io/docs/insights/new-relic-insights/using-new-relic-query-language/nrql-reference#functions

It is working on `develop` locally though, i'm gonna try it on a local prod build just to check

**Comment by LekoArts** [comment]
I've tried some default starters from us and they have correct redirects + our tests are working. I then tried this new relic site and yeah, I also see locally that it doesn't work. At the moment I don't know exactly why.

I think as a next step it would be great to have a minimal reproduction of the problem as running the docs site takes long and a lot of stuff is going on.
Maybe the theme needs updating, maybe it's how the redirects are created (you create redirects both for URLs with and without trailing slash).

**Comment by rudouglas** [comment]
Thanks for looking at that @LekoArts appreciate it, we're going to deprioritise this work for the next few weeks to focus on other things, but when we come back to it we'll keep you updated 🥳

**Comment by jpvajda** [comment]
@rudouglas remind me, what are the next steps on this to resolve it?

**Comment by rudouglas** [comment]
@jpvajda we still haven't discovered a resolution for this so we'll have to do more investigating and testing on it to move it forward :/

**Comment by austin-schaefer** [comment]
Hey @roadlittledawn , let's look at this for a sustaining ticket in the next sprint or two.

**Comment by jpvajda** [comment]
We have a sustaining epic planned in January to kick off Q4, this is a good candidate for that epic, so added it.

**Comment by roadlittledawn** [comment]
looks like this was fixed in gatsby v3.13, so upgrading to this may fix this? https://www.gatsbyjs.com/docs/reference/release-notes/v3.13/#sharp-v029

**Comment by jpvajda** [comment]
We are planning an upgrade to gatsby 4 for the docs site in our sustaining sprint, that should take care of this I believe.

**Comment by rudouglas** [comment]
@roadlittledawn I did try that when it was initially released but it didn't seem to fix the issue

**Comment by jpvajda** [comment]
closing for a nice fresh issue out of the oven 🍪   as this has so much history, we'll work from a new issue to resolve this soon

**Comment by jpvajda** [comment]
https://github.com/newrelic/docs-website/issues/5163
